### DISCUSSION TITLE
Possible use of frameworks for the front-end

### DISCUSSION THREAD

**Discussion by busybox11** [busybox11/NowPlaying-for-Spotify#76] [General]
I've been thinking about rewriting the UI's code for quite some time now, but wasn't really able to do anything due to a lack of time that I could allow to my personal projects.

I now have more time and experience with reactive frameworks such as React, AlpineJS, and front-end frameworks, for instance TailwindCSS. This project was started with poor knowledge of PHP and basically no knowledge of JS whatsoever. I just hacked things together and kept maintaining this (admittedly) hot mess like this.

As you might have guessed, those are my favorite frameworks and I've been reducing my development time on other projects with them. Though, NPfS being more of a lightweight project, I'm concerned about the performance and compatibility issues that they could bring.

I like the idea of NPfS being used as a responsive as-light-as-possible simple visualizer, without any performance overhead. TailwindCSS wouldn't be much of a hassle since it's basically straight up CSS with HTML classes, but some people might disagree with this opinion - that's why I'm making this post. TailwindCSS sounds like a great fit for the project, at least according to me.

About the Reactive framework stuff, even though they might bring some performance overhead and compatibility issues, I would have to test this in the real world with base projects to compare that. I like the simplicity of Alpine, and would guess it's lighter than React, but it's not as known and loved. Going React for this project feels like filling a glass of water with a shower head, but I might be wrong.
Those are my two options. I don't think I'm going to use anything else since I don't have as much experience with other frameworks as I have with those two - but any suggestions are welcome nonetheless.

### TL;DR
Should I rewrite the project with Reactive frameworks and TailwindCSS?

Feel free to post your opinion and debate about this - any opinion is welcome, I would love to hear about it and try to please everyone as much as I can :)

**Comment by ChristianVermeulen** [comment]
I initially posted this in Discord, but I guess this is the better place :).

We have started using Svelte as frontend framework. It is very lightweight and has a not so high learning curve (compared to react). It works pretty intuitively and is set up easily.

https://svelte.dev/

**Reply by busybox11** [reply]
I didn't really know where to post this, both is great actually :)

Thanks for your suggestion! I took a look at it and it does indeed sound pretty cool. I'm really interested by the animations feature they provide. Though the syntax seems a little more verbose than Alpine's, it might be a great lightweight replacement for Vue. It's not as known and used as the other ones but if it's simple enough for most people to understand what's going on, I might consider it.

One drawback would be the fact that it has to be compiled and that would be harder to distribute, as this project is used a lot on simple free PHP server hosting, that is customized by the users. I'm not a fan of having to build the website, but it would be a huge advantage performance-wise.

Thank you a lot for your contribution to this project! :)

**Reply by ChristianVermeulen** [reply]
You could maybe include a ready-built version in the website repo as well :) Most JS packages have both their src ánd built version shipped so the user has the choice of including the minified build or build it with rest of their app using the source.

Another option would be to look into making it a 100% front-end app if we can manage to do the authentication in Svelte as well. This would (i tink?) completely eliminate the use of PHP (or even Node) as backend.

A quick search led me to this oAuth library in Svelte for example: https://github.com/MacFJA/svelte-oauth2

What do you think about that idea?

**Reply by ChristianVermeulen** [reply]
BTW, I just came across this interesting statistic from State of Javascript. Apparently, Svelte is already the #2 framework out there, surpassing Vue, Angular ánd Ember since 2019! https://2019.stateofjs.com/front-end-frameworks/

Looking at their last report, it has only gone up from there :-p
https://2022.stateofjs.com/en-US/libraries/front-end-frameworks/

Anyway, feel free to choose as you see fit, just thought it was interesting to share :-).
I'm in no way a Svelte ambassador or maximalist, just a happy user of it :-p

**Reply by busybox11** [reply]
> You could maybe include a ready-built version in the website repo as well :)

Yeah, I thought about this but it felt wrong somehow. I'm probably going to use TailwindCSS though, which will require compiling / minifying the code anyway, so that's an option.

> Another option would be to look into making it a 100% front-end app if we can manage to do the authentication in Svelte as well. This would (i tink?) completely eliminate the use of PHP (or even Node) as backend.

This was probably the first thing I tried while making this project, but I don't want to expose my API keys. AFAIR it was not possible to do it properly client-side, at the time at least.
If that's possible now, it wouldn't be hard to implement anyway. I'm adding this to my todo list.

**Reply by busybox11** [reply]
I've heard a lot about Svelte lately indeed, and I do remember seeing it on top frameworks charts! I also think Apple Music is rewriting their web client in Svelte.

I'm probably going to choose something like Alpine or Svelte. Both sound like a pretty fitting choice, and their animation features will come very handy! I'll keep you updated!
